### ISSUE TITLE
"Flashing Xiaomi Redmi 9T: Limitations with Multiple Rawprogram Files and Sparse Partition Issues"

### ISSUE THREAD

**Issue by basc0hex** [bkerler/edl#541]
While attempting to flash a Xiaomi device (Redmi 9T) with the latest firmware, which contains multiple rawprogram files (0 to 5) along with the necessary patches, I encountered an unexpected limitation with the flashing tool. The tool does not support more than one rawprogram and patch simultaneously. Despite trying various command structures, I was unable to overcome this limitation.

To work around this issue, I attempted to flash each rawprogram file (X) along with its corresponding patch (X) without closing the device connection with the Sahara and proceeded to flash the subsequent rawprogram files sequentially. However, I encountered a failure during the flashing process of the first rawprogram file (0) due to the presence of sparse partitions in the XML file defining the rawprogram. Below is the XML definition of rawprogram0:

<?xml version="1.0" ?>
<data>
  
  
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="ssd" num_partition_sectors="2" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="8.0" sparse="false" start_byte_hex="0x6000" start_sector="6"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="vbmeta_system.img" label="vbmeta_system" num_partition_sectors="16" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="64.0" sparse="false" start_byte_hex="0x8000" start_sector="8"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="keystore" num_partition_sectors="128" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="512.0" sparse="false" start_byte_hex="0x18000" start_sector="24"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="frp" num_partition_sectors="128" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="512.0" sparse="false" start_byte_hex="0x98000" start_sector="152"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="ffu" num_partition_sectors="2048" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="8192.0" sparse="false" start_byte_hex="0x118000" start_sector="280"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="metadata" num_partition_sectors="4096" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="16384.0" sparse="true" start_byte_hex="0x918000" start_sector="2328"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="persist" num_partition_sectors="8192" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="32768.0" sparse="false" start_byte_hex="0x1918000" start_sector="6424"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="" label="persistbak" num_partition_sectors="8192" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="32768.0" sparse="false" start_byte_hex="0x3918000" start_sector="14616"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="recovery.img" label="recovery" num_partition_sectors="32768" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="131072.0" sparse="" start_byte_hex="0x5918000" start_sector="22808"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="super.img" label="super" num_partition_sectors="2097152" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="8388608.0" sparse="true" start_byte_hex="0xd918000" start_sector="55576"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="userdata.img" label="userdata" num_partition_sectors="0" partofsingleimage="false" physical_partition_number="0" readbackverify="false" size_in_KB="0" sparse="true" start_byte_hex="0x20d918000" start_sector="2152728"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="gpt_main0.bin" label="PrimaryGPT" num_partition_sectors="6" partofsingleimage="true" physical_partition_number="0" readbackverify="false" size_in_KB="24.0" sparse="false" start_byte_hex="0x0" start_sector="0"/>
  <program SECTOR_SIZE_IN_BYTES="4096" file_sector_offset="0" filename="gpt_backup0.bin" label="BackupGPT" num_partition_sectors="5" partofsingleimage="true" physical_partition_number="0" readbackverify="false" size_in_KB="20.0" sparse="false" start_byte_hex="(4096*NUM_DISK_SECTORS)-20480." start_sector="NUM_DISK_SECTORS-5."/>
</data>

Furthermore, the gpt_mainX.bin (X=0 to 5) files belong to the PrimaryGPT partition, and the gpt_backupX.bin (X=0 to 5) files belong to the BackupGPT partition, as defined in different rawprogram files.

Additional Information:
Other sparsed partitions existing in different rawprogram files need to be addressed accordingly to ensure successful flashing.

**Comment by David112x** [comment]
The tool doesn't need to flash more than one rawprogram.xml and patch.xml, the reason why is because the files (0 to 5) are for your device's LUNs in it's UFS, for example, rawprogram0.xml along with patch0.xml only change LUN 0, while rawprogram1.xml and patch1.xml only change LUN 1, this means that after using the command to flash the firmware with the XML files for LUN 0, you can use the same command but instead of using the same XML files for LUN 0, you use the ones for LUN 1 and so on, until you get to the ones for LUN 5, then your device should have everything.

As for the errors regarding the sparse images, your best bet is to unsparse them (probably super.img and userdata.img, because I faced this exact issue on my Poco X3 NFC while flashing it over EDL) and set "sparse" to false for super and userdata in the rawprogram0.xml, after that, you'll have to turn both of them into raw images, you can use `simg2img` to do this.

As janky as it might be to flash a raw super.img and raw userdata.img over EDL, I don't know if there's any other way to flash those two while being sparse images.

TL;DR
Flash with those rawprogram and patch files as well, don't stop at 0.

**Comment by David112x** [comment]
The sparse image issue may apply to other images, but I only remember it happening to me with super.img and userdata.img

**Comment by basc0hex** [comment]
> The tool doesn't need support to flash more than one rawprogram.xml and patch.xml, because you can do it yourself and the reason why is because the files (0 to 5) are there is because of your device's LUNs in it's UFS, for example, rawprogram0.xml along with patch0.xml only change LUN 0, while rawprogram1.xml and patch1.xml only change LUN 1, this means that after using the command to flash the firmware with the XML files for LUN 0, you can use the same command but instead of using the same XML files for LUN 0, you use the ones for LUN 1 and so on, until you get to the ones for LUN 5, then your device should have everything.
> 
> As for the errors regarding the sparse images, your best bet is to unsparse them (probably super.img and userdata.img, because I faced this exact issue on my Poco X3 NFC while flashing it over EDL) and set "sparse" to false for super and userdata in the rawprogram0.xml, after that, you'll have to turn both of them into raw images, you can use `simg2img` to do this.
> 
> As janky as it might be to flash a raw super.img and raw userdata.img over EDL, I don't know if there's any other way to flash those two while being sparse images.
> 
> TL;DR Flash with those rawprogram and patch files as well, don't stop at 0 and make super.img and userdata.img raw images rather than sparse.

i know and i already try that except trying to merge the sparsed files and unsparse them  ... i thought that edl automaticly track the hex addresses from the rawprogram .. anyways thatnks for the answer and sorry for the late answer ^^
